docs/configuration.rst
.. _configuration:

Configuration
=============

The configuration is stored in the ``config.ini`` file.

+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| Section            | Setting           | Default value         | Note                                                            |
+====================+===================+=======================+=================================================================+
| ``context-broker`` | ``host``          | 0.0.0.0               | IP address to accept requests.                                  |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``context-broker`` | ``port``          | 5000                  | TCP port to accept requests.                                    |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``context-broker`` | ``https``         | false                 | Accept only HTTPS requests.                                     |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``auth``           | ``enabled``       | true                  | Enable JWT authentication.                                      |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``auth``           | ``header-prefix`` | GUARD                 | Header prefix for JWT authentication.                           |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``auth``           | ``secret-key``    | guard-secret-key      | Secret key for JWT authentication.                              |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``heartbeat``      | ``timeout``       | 10s                   | Timeout for heartbeat procedure with LCPs.                      |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``heartbeat``      | ``period``        | 1min                  | Period to execute the heartbeat procedure with LCPs.            |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``elasticsearch``  | ``endpoint``      | localhost:9200        | Endpoint connection to Elasticsearch instance.                  |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``elasticsearch``  | ``timeout``       | 20s                   | Timeout for connection to Elasticsearch instance.               |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``elasticsearch``  | ``retry-period``  | 1min                  | Time to wait to retry the connection to Elasticsearch instance. |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``elastic-apm``    | ``enabled``       | false                 | Enable Elastic APM integration.                                 |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``elastic-apm``    | ``server``        | http://localhost:8200 | Elastic APM server.                                             |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+
| ``log``            | ``config``        | log.yaml              | Path of the log configuration file.                             |
+--------------------+-------------------+-----------------------+-----------------------------------------------------------------+


CHANGELOG.md
# Changelog

## 1.0.0

- Initial release.
- Documentation in readthedocs.com (75%).
- Added XML and YAML content type support.
- Added XML and YAML parameter schema.

## 1.0.1

- Fix error in selecting fields.

## 1.0.2

- Fix error to get exact match in query.
- Improved execution of action as daemon.
- Force execution of request even if the LCP operation failed.
- Accept list of strings for eBPF-Program code.
- Added Event management.
- Added Algorithm management.
- Fixed heartbeat period.
- Stdout and stderr response format: plain, lines, and JSON.

## 1.0.3

- Added option to force the use of HTTPS instead of HTTP.
- Added support to enable/disable HTTP Authentication.
- Added pipeline management with CRUD operations.

## 1.0.4

- Replaced HTTP Basic with JWT Authentication.
- Enhanced Log configuration.


reader/config.py
from configparser import BasicInterpolation as Basic_Interpolation
from configparser import ConfigParser as Config_Parser
from os import path
from pathlib import Path


class Config_Reader:
    path = Path(__file__).parent / '../config.ini'

    def __init__(self):
        self.cr = Config_Parser(interpolation=Config_Reader.Env_Interpolation())

    def read(self):
        self.cr.read(self.path.resolve())

        self.cb_host = self.cr.get('context-broker', 'host', fallback='0.0.0.0')
        self.cb_port = self.cr.get('context-broker', 'port', fallback=5000)
        self.cb_https = self.cr.getboolean('context-broker', 'https', fallback=False)

        self.auth = self.cr.getboolean('auth', 'enabled', fallback=True)
        self.auth_header_prefix = self.cr.get('auth', 'header-prefix', fallback='GUARD')
        self.auth_secret_key = self.cr.get('auth', 'secret-key', fallback='guard-secret-key')

        self.hb_timeout = self.cr.get('heartbeat', 'timeout', fallback='10s')
        self.hb_period = self.cr.get('heartbeat', 'period', fallback='1min')

        self.es_endpoint = self.cr.get('elasticsearch', 'endpoint', fallback='localhost:9200')
        self.es_timeout = self.cr.get('elasticsearch', 'timeout', fallback='20s')
        self.es_retry_period = self.cr.get('elasticsearch', 'retry-period', fallback='3min')

        self.elastic_apm_enabled = self.cr.getboolean('elastic-apm', 'enabled', fallback=False)
        self.elastic_apm_server = self.cr.get('elastic-apm', 'server', fallback='http://localhost:8200')

        self.log_config = self.cr.get('log', 'config', fallback='log.yaml')

    def write(self, db):
        # FIXME is it necessary?
        self.cr.set('context-broker', 'port', db.port)
        self.cr.set('elasticsearch', 'endpoint', db.es_endpoint)
        self.cr.set('elasticsearch', 'timeout', db.es_timeout)

        with self.path.open('w') as f:
            self.cr.write(f)

    class Env_Interpolation(Basic_Interpolation):
        """Interpolation which expands environment variables in values."""

        def before_get(self, parser, section, option, value, defaults):
            """Execute before getting the value.

            :param self: class instance
            :param parser: configparser instance
            :param section: section value
            :param option: option value
            :param value: current value
            :param defaults: default values
            :returns: value with expanded variables
            """
            return path.expandvars(value)
